src/chain/shared_state.rs

```diff
@@ -156,6 +157,32 @@ impl SharedState {
                 .collect()
         }
     }
+
+    /// Updates the entry in `their_keys` for `prefix` to the latest known key; if a split
+    /// occurred in the meantime, the keys for sections covering the rest of the address space are
+    /// initialised to the old key that was stored for their common ancestor
+    pub fn update_their_keys(&mut self, prefix: Prefix<XorName>, key: BlsPublicKey) {
+        if let Some(&pfx) = self
+            .their_keys
+            .keys()
+            .find(|pfx| pfx.is_compatible(&prefix))
```

Comment:
say the incoming prefix is `01`, and we have `010` and `011`, will this `compatible` check only make `010` to be updated?
or, we shall only keep one of the `010` or `011`, due to the `neighbouring rule` ?

Reply:
We should actually remove both `010` and `011` in such a case - but this situation can only happen if there is a merge, and we explicitly don't cater to merges for now. I could work on making this algorithm merge-safe, though, if we think it makes sense.

Reply:
Not sure about the plan of `merge` though. maybe @jeanphilippeD will be better to decide whether shall be `merge-safe` or not.

Reply:
I'm OK with having a very short comment here that this will not handle merging prefixes.